Kanban board types

byq77 opened the pull request:
### **User description**
https://github.com/user-attachments/assets/3f9f4a98-6f5b-47ad-9054-2ac4ac675c93

Recently, I've been inspired by the Task Genius extension, which organizes tasks into projects. For the Task Board plugin, I think the obvious equivalent of a project is simply a board with a defined board-level filter.

Following this philosophy, users should be able to switch between different organizations of Kanban columns within the same board. Currently, this requires creating a separate board with a different column organization.

This PR enhances the Task Board plugin by allowing boards to switch between three column organizations: status-based, tag-based, and time-based within the same board. The result should be a more flexible, streamlined experience for organizing and viewing tasks across different contexts.


___

### **PR Type**
Enhancement


___

### **Description**
- Introduces support for multiple Kanban board type organizations (status-based, time-based, tag-based) within the same board

- Refactored board configuration structure to use nested column groups organized by board type instead of flat arrays

- Added `KanbanBoardType` enum with three board type options for type-safe selection

- Implemented settings migration logic to convert legacy board formats to new nested structure seamlessly

- Enhanced board configuration modal with type selector dropdown and board duplication functionality

- Updated all column and task components to use `getActiveColumns()` and `getActiveColumnKey()` helpers for correct column group access

- Added board type state tracking and switching logic in task board view with user notifications

- Expanded language translations (English and Polish) with new keys for board type functionality

- Added CSS styling for board configuration UI improvements including button containers and select element sizing


___

### Diagram Walkthrough


```mermaid
flowchart LR
  A["Board Configuration"] -->|"Select Board Type"| B["KanbanBoardType Enum"]
  B -->|"statusBoard/timeBoard/tagBoard"| C["ColumnGroupData Structure"]
  C -->|"Nested Columns"| D["Active Column Helpers"]
  D -->|"getActiveColumns/getActiveColumnKey"| E["Components & Services"]
  E -->|"Column/TaskItem/KanbanView"| F["Rendered Kanban Board"]
  A -->|"Duplicate Board"| G["Board Duplication"]
  H["Legacy Board Format"] -->|"Migration Logic"| C
```



<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>14 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>en.ts</strong><dd><code>Language translations reorganized and Kanban board type keys added</code></dd></summary>
<hr>

src/utils/lang/locale/en.ts

<ul><li>Reorganized and alphabetically sorted all language translation keys <br>for consistency<br> <li> Added new translation keys: <code>board-type</code>, <code>board-type-info</code>, <code>status-board</code>, <br><code>time-board</code>, <code>tag-board</code>, <code>changed-kanban-board-type</code>, <br><code>no-board-selected-to-duplicate</code>, <code>duplicate-this-board</code>, <code>copy-suffix</code><br> <li> Removed duplicate entries and consolidated related translations<br> <li> Maintained all existing functionality while improving maintainability</ul>


</details>


  </td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-cea8d2b2ef8b629be7f7b64cb086162403b164fbdaa25fb816cdf7829736aad6">+281/-272</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>GlobalSettings.ts</strong><dd><code>Board configuration structure refactored for multiple Kanban types</code></dd></summary>
<hr>

src/interfaces/GlobalSettings.ts

<ul><li>Restructured default board configurations to support multiple Kanban <br>board types (status, time, tag)<br> <li> Changed <code>columns</code> from a flat array to a nested object with <code>status</code>, <br><code>time</code>, and <code>tag</code> properties<br> <li> Added <code>boardType</code> field to board configuration with default value <br><code>KanbanBoardType.statusBoard</code><br> <li> Consolidated three separate default boards into one board with three <br>column organization types</ul>


</details>


  </td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-05cf9666d3f007f096ed0d2136589daea0d9cb3c3d0ea4179a8b137e4d1e65f6">+178/-196</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>SettingSynchronizer.ts</strong><dd><code>Settings migration logic for board structure transformation</code></dd></summary>
<hr>

src/settings/SettingSynchronizer.ts

<ul><li>Added migration logic to convert legacy board format (flat columns <br>array) to new format (nested column groups)<br> <li> Implemented <code>isLegacyBoard()</code> helper to detect old board structure<br> <li> Added <code>resolveDefaultColumns()</code> to extract default column configurations<br> <li> Enhanced migration to handle both legacy and new board formats <br>seamlessly</ul>


</details>


  </td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-2b0edfa363d03f3a9e9ed7dc456b668d859bdb69775a3f082328f90a0ab67f00">+152/-58</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>ColumnSegregator.ts</strong><dd><code>Column segregator updated to support multiple board types</code></dd></summary>
<hr>

src/utils/algorithms/ColumnSegregator.ts

<ul><li>Updated to use <code>getActiveColumns()</code> helper function instead of directly <br>accessing board columns array<br> <li> Refactored named tags filtering to work with active columns from <br>current board type<br> <li> Updated completed column index lookup to use active columns</ul>


</details>


  </td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-93944f139573d036711ef2be76ace8c58bcc5b5658b72bb5a9572cc270db9bb1">+20/-17</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>BoardConfigs.ts</strong><dd><code>Board configuration interfaces refactored with type grouping</code></dd></summary>
<hr>

src/interfaces/BoardConfigs.ts

<ul><li>Added <code>ColumnGroupData</code> type to organize columns by type (status, time, <br>tag)<br> <li> Created <code>BoardLegacy</code> type for backward compatibility with old board <br>format<br> <li> Updated <code>Board</code> type to include <code>boardType</code> field and nested <br><code>ColumnGroupData</code> structure<br> <li> Added helper functions <code>getActiveColumns()</code> and <code>getActiveColumnKey()</code> to <br>retrieve columns based on board type</ul>


</details>


  </td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-c7e767a40c0f6c435e942172d394f58fb5270a38a7ab14e1f6043d7128fa2b2e">+56/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>OpenModals.ts</strong><dd><code>Modal service updated for active column retrieval</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/services/OpenModals.ts

<ul><li>Updated <code>openTaskBoardActionsModal()</code> to use <code>getActiveColumns()</code> helper <br>instead of direct column access<br> <li> Ensures modal receives correct columns based on current board type</ul>


</details>


  </td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-588aaeaddbe4ae280fac47e08068ce931fb46f32e97cd97f8056abd576d7331d">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>Enums.ts</strong><dd><code>Kanban board type enumeration added</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/interfaces/Enums.ts

<ul><li>Added new <code>KanbanBoardType</code> enum with three values: <code>statusBoard</code>, <br><code>timeBoard</code>, <code>tagBoard</code><br> <li> Provides type-safe board type selection for Kanban views</ul>


</details>


  </td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-d9d41c6e79519e28b2cae5fd0776b833ccb6a690aee08cc91f5db7c61cc3d7ff">+6/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>styles.css</strong><dd><code>CSS styling for board configuration UI enhancements</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

styles.css

<ul><li>Added minimum width styling for board config modal select elements<br> <li> Added new button container class <code>boardConfigModalDoubleBtnContainer</code> <br>for side-by-side button layout<br> <li> Added styling for duplicate board button alongside delete button<br> <li> Implemented flex layout for button container with equal distribution</ul>


</details>


  </td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-380b7b38760dd442e897eb0164c58f6a17da966ccaca6318017a468c163979b1">+14/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>BoardConfigModal.tsx</strong><dd><code>Board configuration modal enhanced with type switching and duplication</code></dd></summary>
<hr>

src/modals/BoardConfigModal.tsx

<ul><li>Added board type selector dropdown to allow switching between status, <br>time, and tag-based boards<br> <li> Implemented <code>handleDuplicateCurrentBoard()</code> function to duplicate <br>existing board configurations<br> <li> Updated column manipulation logic to use <code>getActiveColumnKey()</code> for <br>accessing correct column group<br> <li> Modified all column operations (add, delete, reorder) to work with <br>nested column structure<br> <li> Added duplicate and delete buttons in a container for improved UI <br>layout</ul>


</details>


  </td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-59391f26a6b3038949a3f2f7a8cf78b926e07d877144329db2164ee9c68de54d">+84/-23</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>Column.tsx</strong><dd><code>Column component updated for multiple board type support</code>&nbsp; </dd></summary>
<hr>

src/components/KanbanView/Column.tsx

<ul><li>Updated all column operations to use <code>getActiveColumnKey()</code> for <br>accessing correct column group<br> <li> Modified minimize, update, filter, and hide column operations to work <br>with nested column structure<br> <li> Ensured column index lookups use active columns based on board type</ul>


</details>


  </td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-3ca8db3f190debc870f0b24a544a0adf93ef950d844f8dda40d382655fc28992">+18/-10</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>LazyColumn.tsx</strong><dd><code>Lazy column component updated for board type support</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/components/KanbanView/LazyColumn.tsx

<ul><li>Updated lazy-loaded column operations to use <code>getActiveColumnKey()</code> for <br>correct column group access<br> <li> Modified minimize, update, filter operations to work with nested <br>column structure<br> <li> Ensured consistency with Column component changes</ul>


</details>


  </td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-9b6eaa569749bc7e2138a5dbb124c844f083adfbc07c7ac9e8e7ba098a62c879">+18/-10</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>TaskBoardViewContent.tsx</strong><dd><code>Task board view enhanced with board type switching capability</code></dd></summary>
<hr>

src/components/TaskBoardViewContent.tsx

<ul><li>Added <code>activeBoardType</code> state to track current Kanban board type<br> <li> Implemented board type switching logic in view type handler<br> <li> Updated column filtering to use <code>getActiveColumns()</code> helper<br> <li> Added notice message when board type is changed<br> <li> Added dependency on <code>activeBoardType</code> to useMemo hook for proper <br>re-rendering</ul>


</details>


  </td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-4ae1122f5b599f37c089967a9d3e0744d1907022478f345cda3d3ef9c73bd9a4">+21/-10</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>TaskItem.tsx</strong><dd><code>Task item component updated for active column retrieval</code>&nbsp; &nbsp; </dd></summary>
<hr>

src/components/KanbanView/TaskItem.tsx

<ul><li>Updated column data retrieval to use <code>getActiveColumns()</code> helper<br> <li> Ensures task items correctly reference columns from active board type</ul>


</details>


  </td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-6d06ad85bbcdcb63b921f8ced38600799d8dab8d5e6f9e45600f6cce05b676b6">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>KanbanBoardView.tsx</strong><dd><code>Kanban board view updated for active column support</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/components/KanbanView/KanbanBoardView.tsx

<ul><li>Updated to use <code>getActiveColumns()</code> helper for retrieving columns<br> <li> Modified empty board check to use active columns length<br> <li> Ensures Kanban board renders correct columns based on current board <br>type</ul>


</details>


  </td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-71161ae969a3ac09941b254543dcfdbcfe035a929c3706caf44a8ce84195539f">+4/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Localization</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>en.json</strong><dd><code>Alphabetical reorganization and board type localization keys</code></dd></summary>
<hr>

src/utils/lang/locale/en.json

<ul><li>Reorganized locale strings alphabetically for better maintainability<br> <li> Added new localization keys for board type functionality: <code>board-type</code>, <br><code>board-type-info</code>, <code>status-board</code>, <code>tag-board</code>, <code>time-board</code><br> <li> Added key <code>changed-kanban-board-type</code> to track board type changes<br> <li> Preserved all existing translation strings while reordering them</ul>


</details>


  </td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-27b0afabf178bf39753e0c8773e129b691c707f75e829f282d3a2d2b8ec5791c">+437/-426</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>pl.json</strong><dd><code>Comprehensive Polish localization expansion for kanban board types</code></dd></summary>
<hr>

src/utils/lang/locale/pl.json

<ul><li>Expanded Polish language translations from 202 to 482 entries, adding <br>comprehensive support for new features<br> <li> Added translations for kanban board type switching functionality <br>(<code>board-type</code>, <code>status-board</code>, <code>tag-board</code>, <code>time-board</code>)<br> <li> Introduced translations for advanced filtering, sorting, and map view <br>features<br> <li> Added support for new UI elements including board configuration, <br>column management, and task dependencies</ul>


</details>


  </td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-6942fa0c45007359681662d1ecaf61eb98d50d66628a0d2767cdca21794c226e">+455/-175</a></td>

</tr>
</table></details></td></tr><tr><td><strong>Additional files</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>data.json</strong></td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-99377c63fbe5425d01e6e128d8b6a4a9b3f2e18bb233155583875b8a65aef58f">+0/-1004</a></td>

</tr>

<tr>
  <td><strong>tasks.json</strong></td>
  <td><a href="https://github.com/tu2-atmanand/Task-Board/pull/583/files#diff-f6a783534033af7c90bd280e32fc25a49220eb4aaf65b1e07a010a86935ec7ba">+0/-2284</a></td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

tu2-atmanand commented:
Hi @byq77 !

I have been going through this PR and its functionalities from past few days. But didnt got the time to comment here, as was busy with some major bugs which got spawned in the latest enhancements I made for user feature. The latest version has been released just now.

Coming to this PR, you have done exceptional work here. Its really nice to see how you have transformed the plugin for project management. And how cool it looks now. I really liked these new features. I can see the amount of work you have put in here.

Although, these are not very big changes, which are directly affecting the plugins fundamental working. So, I can actually release them in the upcoming version, `1.9.0`. But, I am still taking some time, to think more on this. Because, as you know, after the recent discussion from #561 , since storing board specific data inside dedicated files is a being change in the working of the plugin, hence it will be released in the major version release, `2.0.0`. No timeline is set yet, but its actually a high priority work. 

So, will need to think if, I should release this feature in the upcoming release first. Mostly, I might...

byq77 commented:
I tired make this branch compatible with v1.8.4, but something is broken in latest release, so I force-pushed compatibility back to v1.8.3.

This what I'm getting for v1.8.4...
<img width="1720" height="1408" alt="Screenshot from 2025-12-05 18-16-03" src="https://github.com/user-attachments/assets/29591400-dec0-47d4-b978-2eb4b9a16b6b" />

byq77 commented:
The advantage of my approach is that the Map View associated with the board is not affected by cycling between the Kanban board views. It is pretty obvious to me that users would like to see their tasks in different views on the same board (for example, to see which tasks associated with the board/project are overdue/in progress). The best solution would be to completely disassociate the column organizations from the particular boards and allow users to define multiple column organizations or choose an existing one from a template. I am speaking from the perspective of a user.

tu2-atmanand commented:
Hi @byq77 

I have just released a new version, `1.8.5`. Can you please test it.

Not sure, but everything was working fine in development environment, but its failing when I build the release.
Something wrong with how I used the `useLayoutEffect`. React has its own problems with Obsidian's plugin bundling, or not sure whats happening with the DOM.

Hopefully, it  should fix now, as I reverted the method of rendering.

byq77 commented:
> Hi @byq77
> 
> I have just released a new version, `1.8.5`. Can you please test it.
> 
> Not sure, but everything was working fine in development environment, but its failing when I build the release. Something wrong with how I used the `useLayoutEffect`. React has its own problems with Obsidian's plugin bundling, or not sure whats happening with the DOM.
> 
> Hopefully, it should fix now, as I reverted the method of rendering.

After merging the current main it works fine.

tu2-atmanand commented:
Hi @byq77 !

Just wanted to give one heads-up.

When, I just spin up my development environment, the issue of content vanishing inside the TaskItem card is only showing up in the development environment. But, if I build a production bundle, there is no such issue. Right now, the latest version is using `useEffect` instead of `useLayoutEffect`.

Not, sure what is causing this. Actually, based on AI suggestion, I tried `useLayoutEffect`. But, if use this, then the issue shows up in production bundle, but it doesnt appear in dev environment. Its completely making me crazy. Maybe its just my dev environment, or I am not sure.

I feel, the issue started happening, after I converted the `task.id` from number to string, or somewhere at that time, where now each DOM component is using this as the unique identifier to separate the card components.

## Workaround

One possible reason why this is happening only during development time is, the components are refreshed multiple times, just to have extra safe-checking. And during this multiple rendering, the return statement as mentioned below is cleaning things up.

The workaround includes. Simply comment out the following return statement. Which will avoid cleaning up the content, when the component refreshes again.

https://github.com/tu2-atmanand/Task-Board/blob/aa04a80f37254b147690ebb0d7422e63026b92cf/src/components/KanbanView/TaskItem.tsx#L150-L152

I am not sure, if I should remove it in production build also, whether it will work or not is a mystery, but will see in future.

## More literature

- First time when users started getting this issue : https://github.com/tu2-atmanand/Task-Board/issues/585#issuecomment-3613716358
- After applying `useLayoutEffect` all cards were going blank. Hence, reverted back to `useEffect` : https://github.com/tu2-atmanand/Task-Board/issues/585#issuecomment-3621757620

byq77 commented:
> Hi @byq77 !
> 
> Just wanted to give one heads-up.
> 
> When, I just spin up my development environment, the issue of content vanishing inside the TaskItem card is only showing up in the development environment. But, if I build a production bundle, there is no such issue. Right now, the latest version is using `useEffect` instead of `useLayoutEffect`.
> 
> Not, sure what is causing this. Actually, based on AI suggestion, I tried `useLayoutEffect`. But, if use this, then the issue shows up in production bundle, but it doesnt appear in dev environment. Its completely making me crazy. Maybe its just my dev environment, or I am not sure.
> 
> I feel, the issue started happening, after I converted the `task.id` from number to string, or somewhere at that time, where now each DOM component is using this as the unique identifier to separate the card components.
> 
> ## Workaround
> One possible reason why this is happening only during development time is, the components are refreshed multiple times, just to have extra safe-checking. And during this multiple rendering, the return statement as mentioned below is cleaning things up.
> 
> The workaround includes. Simply comment out the following return statement. Which will avoid cleaning up the content, when the component refreshes again.
> 
> https://github.com/tu2-atmanand/Task-Board/blob/aa04a80f37254b147690ebb0d7422e63026b92cf/src/components/KanbanView/TaskItem.tsx#L150-L152
> 
> I am not sure, if I should remove it in production build also, whether it will work or not is a mystery, but will see in future.
> 
> ## More literature
> * First time when users started getting this issue : [Tasks With Due Dates Lose Their Visible Title on the Board After Completing a Task #585 (comment)](https://github.com/tu2-atmanand/Task-Board/issues/585#issuecomment-3613716358)
> * After applying `useLayoutEffect` all cards were going blank. Hence, reverted back to `useEffect` : [Tasks With Due Dates Lose Their Visible Title on the Board After Completing a Task #585 (comment)](https://github.com/tu2-atmanand/Task-Board/issues/585#issuecomment-3621757620)

Interesting. I did not encounter this (it was showing both when using `npm run dev` and `npm run build`). However, I have developed a practice to always use [devcontainers](https://code.visualstudio.com/docs/devcontainers/containers) for any serious development, sometimes even for embedded development.
